subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyMetadataRulesIntegrationTest.groovy

```diff
@@ -0,0 +1,304 @@
+                        withVariant("$variantToTest") {
+                            withDependencies { dependencies ->
+                                assert dependencies.size() == 1
+                                dependencies.removeAll { true }
+                            }
+                        }
+                    }
+                    all {
+                        withVariant("$variantToTest") { 
+                            withDependencies { dependencies ->
+                                assert dependencies.size() == 0
+                            }
+                        }
+                    }
+                }
+            }
+        """
+
+        then:
+        succeeds 'checkDep'
+        def variantToTest = variantToTest
+        resolve.expectGraph {
+            root(':', ':testproject:') {
+                module("org.test:moduleA:1.0:$variantToTest")
+            }
+        }
+    }
+
+    def "changing dependencies in one variant leaves other variants untouched"() {
+        when:
+        buildFile << """
+            dependencies {
+                components {
+                    withModule('org.test:moduleA') {
+                        withVariant("default") {
+                            withDependencies {
+                                add('org.test:moduleB:1.0')
+                            }
+                        }
+                    }
+                }
+            }
+        """
+
+        then:
+        succeeds 'checkDep'
+        def variantToTest = variantToTest
+        resolve.expectGraph {
+            root(':', ':testproject:') {
+                module("org.test:moduleA:1.0:$variantToTest")
+            }
+        }
+    }
+
+    def "dependencies of transitive dependencies can be changed"() {
+        given:
+        setupCustomVariantsForModule(moduleB).publish()
+        moduleA.dependsOn(moduleB).publish()
+        repo.module("org.test", "moduleC").allowAll().publish() //.withModuleMetadata().variant("anotherVariant", [format: "custom"]).publish()
+
+        when:
+        buildFile << """
+            dependencies {
+                components {
+                    withModule('org.test:moduleB') {
+                        withVariant('$variantToTest') {
+                            withDependencies {
+                                add('org.test:moduleC:1.0')
+                            }
+                        }
+                    }
+                }
+            }
+        """
+
+        then:
+        succeeds 'checkDep'
+        def variantToTest = variantToTest
+        resolve.expectGraph {
+            root(':', ':testproject:') {
+                module("org.test:moduleA:1.0:$variantToTest") {
+                    module("org.test:moduleB:1.0") {
+                        module('org.test:moduleC:1.0')
+                    }
+                }
+            }
+        }
+    }
+
+    def "attribute matching is used for all dependencies added by rules"() {
```

Comment:
⭕️ This is confusing to me (including the comment further down). Matching happens on variants, not on dependencies.

Reply:
I changed the name and also a sentence in the comment.

Basically I put this in here to document this behaviour:
We chose to always create `GradleDependencyMetadata` no matter if the dependency is added to a Gradle, Mvn or Ivy metadata. And the variant selection in the dependency's target depends on the `selectConfigurations()` implementation.